### PR TITLE
Conformance 2.0 Test Pack

### PR DESCRIPTION
**Pull request by robmoffat**

### CODE REVIEW

File: toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md [hunk 1]
```diff
@@ -0,0 +1,399 @@
+# FDC3 2.0 Conformance Test Cases
+
+## 1. Basic Tests
+
+_These are some basic sanity tests implemented in the FDC3 Conformance Framework.  It is expected that Desktop Agent testers will run these first before commencing the much more thorough tests in section 2 onwards._
+
+- `2.0-BasicCL1`: You can call the `fdc3.addContextListener` with the `fdc3.contact` context type. The returned listener object has an `ubsubscribe` function.
+- `2.0-BasicCL2`: You can call the `fdc3addContextListener` with no context type. The returned listener object has an `ubsubscribe` function.
```
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md:8]:
```suggestion
- `2.0-BasicCL2`: You can call the `fdc3.addContextListener` with a `null` context type. The returned listener object has an `unsubscribe` function.
```

File: toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+- `2.0-BasicIL1`: You can call the `fdc3.addIntentListener` on the `DesktopAgent` for an intent, and get back a `Listener` object with `unsubscribe` method.
```
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md:9]: returns a Promise<Listener> in 2.0

File: toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+- `2.0-BasicCH1`: A call to `fdc3.getCurrentChannel` on the `DesktopAgent` always returns a promise.
+- `2.0-BasicCH2`: A call to `fdc3.getCurrentChannel()` returns a promise resolving to _null_ if called prior to any `joinChannel`.
```
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md:11]: Needs correcting to 2.0 function name

File: toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+- `2.0-BasicGI1`: A call to `fdc3.getInfo()` returns an object with `fdc3Version` and `provider` properties.
```
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md:12]: Returns a Promise in 2.0

File: toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+- `2.0-BasicAC1`: A call to `fdc3.getOrCreateChannel(<name>)` will return an object matching the `Channel` interface, with properties of `id`, `type`, `broadcast`, `getCurrentContext` and `addContextListener`.
```
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md:13]: returns a  promise in both 1.2 and 2.0

File: toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+- `2.0-BasicUC1`: You can call the `fdc3.getSystemChannels()` function and receive a promise containing an array of more than 1 `Channel` objects, each with type and id set.
```
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md:14]: update to 2.0 function name

File: toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+- `2.0-BasicJC1`: You can call `fdc3.joinChannel`, passing in the `id` of one of the system channels.  `fdc3.getCurrentChannel` should then return that joined channel.
```
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md:15]: update to 2.0 function name.
async in both 1.2 and 2.0 so should be awaited

File: toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+| A   |1.  addContextListener |Call `fdc3.addContextListener(null, handler)`<br>Check **promise** resolving a **listener**  object is returned<br>Check that there is an `unsubscribe` function on the **listener object**   |
+| A   | 2. joinUserChannel        |`fdc3.getUserChannels()`<br>Check **user** channels are returned.<br>Call `fdc3.joinUserChannel()` on first **user** channel                                      |
+| B   | 3. joinUserChannel        | `fdc3.getUserChannels()`<br>Check **user** channels are returned.<br>Call `fdc3.joinUserChannel()` on first **user** channel                                      |
+| B   | 4. Broadcast          | `fdc3.broadcast(<some instrument>)` <br> **Check promise of `void` is returned**                                                                                                                         |
+| A   |5.  Receive Context    | Instrument object matches the one broadcast in 2 above.                                                                                                    |
+
+- `2.0-BasicUsage1` Perform above test
```
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md:31]: Please improve test naming

File: toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+| A   | 2. joinUserChannel        |`fdc3.getUserChannels()`<br>Check **user** channels are returned.<br>Call `fdc3.joinUserChannel()` on first **user** channel                                      |
+| B   | 3. joinUserChannel        | `fdc3.getUserChannels()`<br>Check **user** channels are returned.<br>Call `fdc3.joinUserChannel()` on first **user** channel                                      |
+| B   | 4. Broadcast          | `fdc3.broadcast(<some instrument>)` <br> `fdc3.broadcast()` a contact context.   <br> **Check promise of `void` is returned in both cases**                 |
+| A   | 5. Receive Context    | Instrument object matches the one broadcast in 2 above.<br>Check that the contact is not received.                                                                   |
+
+- `2.0-FilteredUsage1` Perform above test 
+- `2.0-FilteredUsage2` Perform steps in order: 2,1,3,4,5
+- `2.0-FilteredUsage3` 3,4,1,2,5
+- `2.0-FilteredUsage4` 3,4,2,1,5
```
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md:49]: Please improve test naming

File: toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+| A   | 5. Receive Context    | Instrument object matches the one broadcast in 2 above.<br>Contact object matches the one broadcast in 2 above.                                                                                                                                                                                                                                   |
+
+ - `2.0-UCFilteredUsage5`: Perform above test
+ - `2.0-UCFilteredUsage6`: Perform above test, except B will join a _different_ channel to A. Check that you _don't_ receive anything.
+ - `2.0-UCFilteredUsageUnsubscribe`: Perform above test, except that after joining, **A** then `unsubscribe()`s from the channel using the `listener.unsubscribe` function. Check that **A** _doesn't_ receive anything.
+ - `2.0-UCFilteredUsageChange`: Perform above test, except that after joining, **A** changes channel with a further _different_ channel.  Check that **A** _doesn't_ receive anything.
```
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md:62]:
```suggestion
 - `2.0-UCFilteredUsageChange`: Perform above test, except that after joining, **A** changes channel to a _different_ channel via a further call to `fdc3.joinUserChannel`.  Check that **A** _doesn't_ receive anything.
```

File: toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+ - `2.0-UCFilteredUsageLeave`: Perform above test, except that after joining, **A** calls `fdc3.leaveChannel()` and doesn't receive anything.
```
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md:63]:
```suggestion
 - `2.0-UCFilteredUsageLeave`: Perform above test, except that after joining, **A** calls `fdc3.leaveCurrentChannel()` and doesn't receive anything.
```

File: toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md [hunk 12]
```diff
@@ -0,0 +1,399 @@
...
+
+### Specific Context
+
+| App | Step            | Description                                                                                                                   |
+|-----|-----------------|-------------------------------------------------------------------------------------------------------------------------------|
+| A   | Opening App     | Open method as in `AOpensBWithContext` but pass a context type that app B is NOT going to add a listener for (including NOT adding a listener for type `null`)                                                                |
+| B   | Context not received | App B calls `fdc3.addContextListener(type, handler)` for the type's it is listening for but NOT for the type A is sending in this case. Confirm that NO context is received.                         |
+| A   | Promise         | Receives a rejection from the open promise with “App Timeout’ from <br>https://fdc3.finos.org/docs/api/ref/Errors#openerror |
```
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md:149]:
```suggestion
| A   | Promise         | Receives a rejection from the open promise with `AppTimeout` from <br>https://fdc3.finos.org/docs/api/ref/Errors#openerror |
```

File: toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md [hunk 13]
```diff
@@ -0,0 +1,399 @@
...
+
+- `2.0-AOpensBWithWrongContext`: Perform above test
+- `2.0-AOpensBNoListen`: Skip `fdc3.addContextListener()` above (i.e. no listener was added).
```
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md:152]: Define expected result (AppTimeout again)

**Review by kriswest — CHANGES_REQUESTED** [review]
Partial review highlighting a number of corrections needed

**Review by nkolba — APPROVED** [review]
lgtm

File: toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md [hunk 14]
```diff
@@ -0,0 +1,401 @@
...
+| Test   | 3. IntentResolution          | The `raiseIntent` call returns an `IntentResolution` Object with an `AppIdentifer` as the `source field` with App B's `appId` and `instanceId` set.                     |
+| Test   | 4. await results          | Test should `await resolution.getResult()` on the `IntentResolution` object returned in the previous step. A promise should be returned quickly.                        |
+| B       | 5. return void          | B should return void after a short delay (e.g. 5 seconds).                        |
+| Test   | 6. receive void result          | The promise received by Test from `resolution.getResult()` should resolve to void. Confirm that the promise could be retrieved before the handler function returned and that the result was received *after* the result was returned by B, not before. I.e. confirm that `resolution.getResult() does not block until the result is returned, but rather returns a promise that can be awaited.`                      |
```
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md:326]: This one's description needs improving (the await was making the test impossible to implement):

```suggestion
| Test   | 4. await results          | Test should `const resultPromise = resolution.getResult()` on the `IntentResolution` object returned in the previous step. A promise should be returned quickly.                        |
| B       | 5. return void          | B should return void after a short delay (e.g. 5 seconds).                        |
| Test   | 6. receive void result          | The promise received by Test from `resolution.getResult()` should eventually resolve to void and not be rejected. Additionally, confirm that the promise could be retrieved before the handler function returns. I.e. confirm that `resolution.getResult()` does not block until the result is returned, but rather returns a promise that can be awaited.`     |
```

**Review by kriswest — CHANGES_REQUESTED** [review]
A few wording improvements and a correction to code snippet in 'Raise Intent Result (void result)' steps

File: toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md [hunk 15]
```diff
@@ -0,0 +1,408 @@
...
+  - Before the first step, use `let appIdentifier = await fdc3.open({appId: "<A's appId>"})` to start A.
+  - Then use `const instances = await fdc3.findInstances({appId: "<A's appId>"})` to retrieve a list of instances of app A. Confirm that only one is present and retrieve its `AppIdentifier`, confirming that it contains an `instanceId` field that matches that returned by the `fdc3.open` call.
+  - Then in the first step, use `fdc3.raiseIntent("aTestingIntent", testContextX, appIdentifier)` to target the running instance of app A. 
+  - Confirm that the intent is delivered to the correct instance and that another instance is NOT started. Otherwise, as above.
+- `2.0-RaiseIntentFailedResolve`: Perform above test, but:
+  - Use `fdc3.raiseIntent("aTestingIntent", testContextY)`.  Note that no app supports this intent and context combination.**
+  - You should receive a JavaScript Error with the message `ResolveError.NoAppsFound`.
+- `2.0-RaiseIntentFailTargetedAppResolve1`: Perform above test, but:
+  - Use `fdc3.raiseIntent("aTestingIntent", testContextY, {appId: "<A's appId>"})`.
+  - You should receive a JavaScript Error with the message `ResolveError.NoAppsFound`.
```
Comment (kriswest) [inline toolbox/fdc3-conformance/FDC3-2.0-Conformance-Test-Cases.md:303]: Wrong error specified
```suggestion
  - You should receive a JavaScript Error with the message `ResolveError.IntentDeliveryFailed`.
```

**Pull request closed without merging**